templates/usr/local/sbin/simp_generate_types.epp

```diff
@@ -3,132 +3,259 @@
+  @logfile.puts(msg.join("\n")) if @logfile
+
+  msg.each do |line|
+    LOG.error(line) if options.syslog
+  end
+end
+
+def process_paths(to_process, options)
+  to_process.map!(&:strip)
+
+  target_environments = []
+
+  environment_paths = Dir.glob(
+    %x{#{PUPPET_CMD} config print --log_level=err --section=master environmentpath}.strip.split(':').map{|x| x + '/*'}
+  )
+
+  if to_process.any?{|path| path.casecmp('all')==0}
+    target_environments = environment_paths.compact
+  else
+    to_process.each do |path|
+      if environment_paths.include?(path) || environment_paths.any?{|env_path| path.start_with?("#{env_path}/")}
+        target_environments << path
+      end
     end
+
+    to_process = to_process.sort.uniq
+  end
+
+  if target_environments.empty?
+    msg = 'Error: Could not find any target environments'
+    log_err(msg, options)
+    exit 1
+  end
+
+  # Convert each discovered environment into a hash noting the path and environment name
+  target_environments.map!{|e| { :path => e, :env_name => File.basename(e) }}
+
+  output = []
+
+  current_umask = File.umask
+  File.umask(0022)
+
+  puppet_group = %x(#{PUPPET_CMD} config print --log_level=err --section=master group).strip
+
+  target_environments.each do |env|
+    begin
+      log("Generating Types for #{env[:env_name]}", options)
+
+      output += %x{#{PUPPET_CMD} generate types --log_level=err --environment #{env[:env_name]} 2>&1}.strip.lines
+    ensure
+      resource_types_dir = File.join(env[:path], '.resource_types')
+
+      if File.exist?(resource_types_dir)
+        FileUtils.chown_R(nil, puppet_group, resource_types_dir)
+        FileUtils.chmod_R('g+r,o-rwx', resource_types_dir)
+        FileUtils.chmod('g+x', resource_types_dir)
+      end
+    end
+
+    # Give each environment a couple of seconds to process to reduce load on the server
+    sleep(2)
+  end
+
+  output = output.compact.uniq.map(&:strip)
+  output.select!{|x| x.include?('Error:')}
+
+  File.umask(current_umask)
+
+  unless output.empty?
+    log_err(output, options)
+    exit 2
   end
 end
 
-output = output.compact.uniq.map(&:strip)
-output.select!{|x| x.include?('Error:')}
+unless options.path
+  msg = "Error: You must pass a path"
+
+  log_err([msg, options.help], options)
 
-File.umask(current_umask)
+  exit 1
+end
 
-unless output.empty?
-  $stderr.puts output.join("\n") unless options.quiet
+if options.message
+  wait_and_lock(WRITE_LOCK)
+
+  begin
+    top_dir = File.dirname(options.path)
 
-  fh.puts(output.join("\n")) if fh
+    FileUtils.mkdir_p(top_dir) unless File.exist?(top_dir)
 
-  output.each do |line|
-    log.error(line) if options.syslog
+    File.open(options.path, 'a+', 0640) { |fh|
+      fh.puts(options.message)
+      fh.flush
+    }
+  rescue => e
+    LOG.error("Could not open file '#{options.path}' for writing: #{e}")
   end
+elsif options.generate
+  sleep(options.delay)
 
-  exit 2
+  wait_and_lock(WRITE_LOCK)
+
+  if File.exist?(options.path)
```

Comment:
Shouldn't this be before the 'wait_and_lock(WRITE_LOCK)'?

Reply:
Yep